### ISSUE TITLE
fabric failed to save optimizer

### ISSUE THREAD

**Issue by hiyyg** [Lightning-AI/lightning#18688]
### Bug description

```
_pickle.PicklingError: Can't pickle <class 'lightning.fabric.wrappers.FabricRanger'>: attribute lookup FabricRanger on lightning.fabric.wrappers failed
```

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
I followed this way to save a Ranger optimizer.

# Define the state of your program/loop
state = {"model1": model1, "model2": model2, "optimizer": optimizer, "iteration": iteration, "hparams": ...}

fabric.save("path/to/checkpoint.ckpt", state)

```
```


### Error messages and logs

```
# Error messages and logs here please
_pickle.PicklingError: Can't pickle <class 'lightning.fabric.wrappers.FabricRanger'>: attribute lookup FabricRanger on lightning.fabric.wrappers failed
```


### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>


### More info

_No response_

**Comment by hiyyg** [comment]
The same for AdamW:
```
_pickle.PicklingError: Can't pickle <class 'lightning.fabric.wrappers.FabricAdamW'>: attribute lookup FabricAdamW on lightning.fabric.wrappers failed
```

**Comment by hiyyg** [comment]
@awaelchli This is not an error with Ranger. The problem is that I defined an lr scheduler **after** `fabric.setup(model, optimizer)`. If I define the scheduler before `fabric.setup`, the error will be gone.

**Comment by hiyyg** [comment]
Another problem is that, if the scheduler is defined by `torch.optim.lr_scheduler.LambdaLR`, such as 
```
scheduler = torch.optim.lr_scheduler.LambdaLR(optimizer, lambda x: (1 - x / 1000) ** 0.9)

state = {"scheduler": scheduler}
```
trying to `fabric.save('state.pt', state)` will raise the following error:
```
AttributeError: Can't pickle local object 'main.<locals>.<lambda>'
```

So I have to change the code like:
```
state = {"scheduler": scheduler.state_dict()}
fabric.save('state.pt', state)

fabric.load("state.pt", state)
scheduler.load_state_dict(state['scheduler'])
```
In this way, I have to manually do `load_state_dict`, but it seems this is the only way to bypass the error.

I wonder whether it is always safe to do save/load like this for all objects that have `.state_dict()` and `.load_state_dict()`? Because directly saving the objects is buggy now. Also mixing objects and state_dicts is not a good idea.

**Comment by hiyyg** [comment]
It seems another way is to use:
```
state = {"scheduler": scheduler}
save_state = {"scheduler": scheduler.state_dict()}
fabric.save('state.pt', save_state)

fabric.load("state.pt", state)
```
Could this way safely load the state_dicts to the objects?

**Comment by hiyyg** [comment]
Maybe the code here https://github.com/Lightning-AI/lightning/blob/a49092fefc9c579dce393b45219bb4943716bb7d/examples/fabric/image_classifier/train_fabric.py#L102 should be considered to move before `setup`.

**Comment by awaelchli** [comment]
Yes, you'd have to install from source to get this feature. But it's only a nice-to-have. As I wrote, and as you figured out yourself, doing the `.state_dict()` and `.load_state_dict()` yourself can also be done and would be equivalent (you'd do this in raw PyTorch too). 

So feel free to install from source and test it out, but it's not mandatory :)

**Comment by hiyyg** [comment]
> It seems another way is to use:
> 
> ```
> state = {"scheduler": scheduler}
> save_state = {"scheduler": scheduler.state_dict()}
> fabric.save('state.pt', save_state)
> 
> fabric.load("state.pt", state)
> ```
> 
> Could this way safely load the state_dicts to the objects?

Thanks.

It seems the above method can remove the need for `.load_state_dict()`, but need to manually do `.state_dict()` just for saving.
